### ISSUE TITLE
ansible_managed: {file} expands to filename only

### ISSUE THREAD

**Issue by null-git** [ansible/ansible#75321]
### Summary

When I use  `{{ ansible_managed }}` in a template the string variable `{file}` isn't expanded to the full template path anymore but only to the file name.

https://docs.ansible.com/ansible/2.4/intro_configuration.html#ansible-managed states that `{file}` should still be replaced by the full path.

### Issue Type

Bug Report

### Component Name

template

### Ansible Version

```console
$ ansible --version
ansible [core 2.11.2] 
  config file = /usr/local/etc/ansible/ansible.cfg
  configured module search path = ['/usr/local/etc/ansible/modules']
  ansible python module location = /usr/local/lib/python3.8/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/local/share/py38-ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.8.11 (default, Jul 19 2021, 08:36:04) [Clang 11.0.1 (git@github.com:llvm/llvm-project.git llvmorg-11.0.1-0-g43ff75f2c
  jinja version = 2.11.2
  libyaml = False
```

### Configuration

```console
$ ansible-config dump --only-changed
DEFAULT_FORKS(/usr/local/etc/ansible/ansible.cfg) = 10
DEFAULT_HOST_LIST(/usr/local/etc/ansible/ansible.cfg) = ['/usr/local/etc/ansible/inventory']
DEFAULT_LOG_PATH(/usr/local/etc/ansible/ansible.cfg) = /var/log/ansible.log
DEFAULT_MANAGED_STR(/usr/local/etc/ansible/ansible.cfg) = {file}
DEFAULT_MODULE_PATH(/usr/local/etc/ansible/ansible.cfg) = ['/usr/local/etc/ansible/modules']
DEFAULT_NO_TARGET_SYSLOG(/usr/local/etc/ansible/ansible.cfg) = True
DEFAULT_POLL_INTERVAL(/usr/local/etc/ansible/ansible.cfg) = 5
DEFAULT_ROLES_PATH(/usr/local/etc/ansible/ansible.cfg) = ['/usr/local/etc/ansible/roles']
DEFAULT_TRANSPORT(/usr/local/etc/ansible/ansible.cfg) = jail
INVENTORY_IGNORE_EXTS(/usr/local/etc/ansible/ansible.cfg) = ['.md']
RETRY_FILES_ENABLED(/usr/local/etc/ansible/ansible.cfg) = False
```

### OS / Environment

FreeBSD 13.0-p3

### Steps to Reproduce

/usr/local/etc/ansible/roles/test/templates/file.ext:
```
# {{ ansible_managed }}
```

### Expected Results

I expect this to expand to:
```
# /usr/local/etc/ansible/roles/test/templates/file.ext
```

### Actual Results

```console
# file.ext
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by samdoran** [comment]
I believe this has been the case since 2.4.0, introduced here 5b7e8a7bd65f1308e58c3f39627e78db56a3ca39. Did you just notice this recently?

Seems we just need to use the fullpath in the managed string.

```patch
diff --git lib/ansible/template/__init__.py lib/ansible/template/__init__.py
index 33e304ff66..edd05d0286 100644
--- lib/ansible/template/__init__.py
+++ lib/ansible/template/__init__.py
@@ -139,7 +139,7 @@ def generate_ansible_template_vars(path, fullpath=None, dest_path=None):
     managed_str = managed_default.format(
         host=temp_vars['template_host'],
         uid=temp_vars['template_uid'],
-        file=temp_vars['template_path'],
+        file=temp_vars['template_fullpath'],
     )
     temp_vars['ansible_managed'] = to_text(time.strftime(to_native(managed_str), time.localtime(os.path.getmtime(b_path))))
 
```

**Comment by bcoca** [comment]
no, this was done on purpose since 2 users executing 'from' different repository checkouts would get diff pahts `~/bcoca` vs `~/sdoran/projexts/stuff/templates/file.j2`. If you must have the full path, you can use `playbookd_dir` or `role_path` variables

**Comment by samdoran** [comment]
That makes sense. I thought `{file}` was the absolute destination path, but I see why the old behavior was problematic. Having the `ansible_managed` string constantly changing by default is not good.

It looks like we removed descriptions of the `{host}`, `{uid}`, and `{file}` format strings from the docs, though they are demonstrated [here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#adding-comments-to-files) now.

**Comment by bcoca** [comment]
`template_fullpath` has the full path if you need it, its jut not a good default for `ansible_managed`

**Comment by samdoran** [comment]
@null-git Another option is using the `template_` variables instead of or in addition to `ansible-managed`.

> Additional variables listed below can be used  in templates. `ansible_managed` (configurable via the `defaults` section of `ansible.cfg`) contains a string which can be used to  describe the template name, host, modification time of the template file and the owner uid. `template_host` contains the node name of  the template's machine. `template_uid` is the numeric user id of the owner. `template_path` is the path of the template. `template_fullpath` is the absolute path of the template. `template_destpath` is the path of the template on the remote system (added in 2.8). `template_run_date` is the date that the template was rendered.

**Comment by bcoca** [comment]
or just define it as a normal template `ansible_managed: 'file: {{ template_fullpath }}'`

**Comment by null-git** [comment]
@samdoran
> I believe this has been the case since 2.4.0, introduced here [5b7e8a7](https://github.com/ansible/ansible/commit/5b7e8a7bd65f1308e58c3f39627e78db56a3ca39). Did you just notice this recently?

This was a recent change in the FreeBSD package. The last server I've set up with Ansible in February this year still has absolute filepaths in the header. But in February the package was still Ansible 2.9.7 (no idea which core version this is, sorry; https://www.freshports.org/sysutils/ansible/ has the FreeBSD version history at the bottom).
Now after updating some servers to FreeBSD 13 and rerunning Ansible I got a lot of changed messages so I investigated.

@bcoca
> no, this was done on purpose since 2 users executing 'from' different repository checkouts would get diff pahts ~/bcoca vs ~/sdoran/projexts/stuff/templates/file.j2. If you must have the full path, you can use playbookd_dir or role_path variables

But just having the filename makes this rather useless, doesn't it? I use `{file}` to help me find the template for a file and I imagine other Ansible users do / did the same. I've never used `{{ ansible_managed }}` directly though but have a variable in my inventory that removes everything to my roles directory so that I just get `<role_name>/templates/<template_name>`

> or just define it as a normal template ansible_managed: 'file: {{ template_fullpath }}'
Thanks, that fixed the problem for me.


As this was an intended change the issue is now about the wrong documentation:
https://docs.ansible.com/ansible/2.4/intro_configuration.html#ansible-managed :
> `{file}`: expands to the name of the full path to the template file.
